File: JavaScript/116.填充每个节点的下一个右侧节点指针.js
/*
 * @lc app=leetcode.cn id=116 lang=javascript
 *
 * [116] 填充每个节点的下一个右侧节点指针
 */

// @lc code=start
/**
 * // Definition for a Node.
 * function Node(val, left, right, next) {
 *    this.val = val === undefined ? null : val;
 *    this.left = left === undefined ? null : left;
 *    this.right = right === undefined ? null : right;
 *    this.next = next === undefined ? null : next;
 * };
 */

/**
 * @param {Node} root
 * @return {Node}
 */
// [1,2,3,4,5,6,7]
var connect = function (root) {
  const res = []
  const queue = []
  queue.push(root)
  while(queue.length) {
    const node = queue.shift()
    res.push([node])
    let len = queue.length
    while(len) {
      res[res.length - 1]
      queue.push()
      len--
    }
  }
  return res
};
// @lc code=end

File: Python3/116.填充每个节点的下一个右侧节点指针.py
#
# @lc app=leetcode.cn id=116 lang=python3
#
# [116] 填充每个节点的下一个右侧节点指针
#

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

"""
感觉这种题，就是脑筋急转弯，没啥意思。
Accepted
58/58 cases passed (172 ms)
Your runtime beats 8.15 % of python3 submissions
Your memory usage beats 73.8 % of python3 submissions (16.6 MB)
"""


class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root

        def tool(left, right):
            if not left and not right:
                return
            left.next = right
            tool(left.left, left.right)
            tool(right.left, right.right)
            tool(left.right, right.left)

        tool(root.left, root.right)
        return root

# @lc code=end
